backend/src/main/java/com/google/sps/servlets/AddParticipantServlet.java

```diff
@@ -56,33 +58,48 @@
 
   @Override
   public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    // Request parameter values
-    // TODO: Check input values are filled (before allowing to submit)
-    UserService userService = UserServiceFactory.getUserService();
-    String email = userService.getCurrentUser().getEmail();
-    if (email == null) {
-      response.sendError(HttpServletResponse.SC_BAD_REQUEST, "Invalid email.");
-      return;
+
+    // Retrieve body of HTTP request
+    StringBuffer requestBuffer = new StringBuffer();
+    try {
+      BufferedReader reader = request.getReader();
+      String currentLine;
+      while ((currentLine = reader.readLine()) != null) {
+        requestBuffer.append(currentLine);
+      }
+    } catch (Exception e) {
+      response.sendError(HttpServletResponse.SC_BAD_REQUEST, "Could not read request body");
     }
-    String username = email.split("@")[0];
+    JSONObject obj = new JSONObject(requestBuffer.toString());
+
+    // Retrieve the timeAvailableUntil input and convert to a UTC ZonedDateTime
+    long timeAvailableUntil = obj.getJSONObject("formDetails").getLong("timeAvailableUntil");
+    ZoneId zoneId = ZoneId.of("UTC");
+    ZonedDateTime startTimeAvailable = ZonedDateTime.now(Clock.systemUTC());
+    ZonedDateTime endTimeAvailable =
+        ZonedDateTime.ofInstant(Instant.ofEpochMilli(timeAvailableUntil), zoneId);
 
-    String timezone = request.getParameter("timezone");
-    ZoneId zoneId = ZoneId.of(timezone); // TODO: convert input timezone to valid ZoneId
-    ZonedDateTime startTimeAvailable =
-        ZonedDateTime.now(zoneId); // TODO: set to future time if not available now
-    Instant endTimeAvailableInstant =
-        Instant.ofEpochMilli(
-            Long.parseLong(
-                request.getParameter("endTimeAvailable"))); // TODO: figure out input format
-    ZonedDateTime endTimeAvailable = endTimeAvailableInstant.atZone(zoneId);
-
-    int duration = convertToPositiveInt(request.getParameter("duration"));
+    int duration = obj.getJSONObject("formDetails").getInt("duration");
     if (duration <= 0) {
       response.sendError(HttpServletResponse.SC_BAD_REQUEST, "Invalid duration.");
       return;
     }
 
-    long timestamp = System.currentTimeMillis();
+    String role = obj.getJSONObject("formDetails").getString("role");
```

Comment:
obj.getJSONObject("fromDetails") could be extracted to a local variable here since it's referenced on four lines consecutively

Reply:
+1

Reply:
Done!